src/ruis/render/scene/gltf_loader.cpp

```diff
@@ -0,0 +1,772 @@
+
+		for (const auto& index : it->second.array())
+			arr.push_back(index.number().to_uint32());
+	}
+	return arr;
+}
+
+template <typename tp_type, size_t dimension>
+r4::vector<tp_type, dimension> read_vec(
+	const jsondom::value& json,
+	const std::string& name,
+	const r4::vector<tp_type, dimension> default_value
+)
+{
+	auto it = json.object().find(name);
+
+	if (it == json.object().end() || !json.object().at(name).is_array())
+		return default_value;
+
+	r4::vector<tp_type, dimension> value;
+	int i = 0;
+	for (const auto& subjson : json.object().at(name).array()) {
+		value[i++] = subjson.number().to_float();
+	}
+	return value;
+}
+
+ruis::quat read_quat(const jsondom::value& json, const std::string& name, const ruis::quat default_value)
+{
+	ruis::vec4 default_vec = default_value.to_vector4();
+	auto vec = read_vec(json, name, default_vec);
+	return ruis::quat{vec};
+}
+
+} // namespace
+
+utki::shared_ref<buffer_view> gltf_loader::read_buffer_view(const jsondom::value& buffer_view_json)
+{
+	const uint32_t byte_length = read_uint(buffer_view_json, "byteLength");
+	const uint32_t byte_offset = read_uint(buffer_view_json, "byteOffset");
+	const uint32_t byte_stride = read_uint(buffer_view_json, "byteStride");
+	const uint32_t target = read_uint(buffer_view_json, "target");
+
+	auto new_buffer_view =
+		utki::make_shared<buffer_view>(byte_length, byte_offset, byte_stride, static_cast<buffer_view::target>(target));
+	return new_buffer_view;
+}
+
+template <typename tp_type>
+void gltf_loader::create_vertex_buffer_float(
+	utki::shared_ref<ruis::render::accessor> new_accessor,
+	utki::span<const uint8_t> buffer,
+	uint32_t acc_count, // in elements (e.g. a whole vec3)
+	uint32_t acc_stride // in bytes
+)
+{
+	utki::deserializer d(buffer);
+	std::vector<tp_type> vertex_attribute_buffer;
+	vertex_attribute_buffer.reserve(acc_count);
+
+	int n_skip_bytes = int(acc_stride) - int(sizeof(tp_type));
+	if (n_skip_bytes < 0)
+		n_skip_bytes = 0;
+
+	for (uint32_t i = 0; i < acc_count; ++i) {
+		tp_type t;
+
+		if constexpr (std::is_same_v<tp_type, float>) {
+			t = d.read_float_le();
+		} else {
+			for (uint32_t j = 0; j < t.size(); ++j)
+				t[j] = d.read_float_le();
+		}
+		// also, only float vectors for now. Need a template<> deserialize function for nice notation
+		vertex_attribute_buffer.push_back(std::move(t));
+		d.skip(n_skip_bytes);
+	}
+
+	new_accessor.get().vbo = factory_v.create_vertex_buffer(utki::make_span(vertex_attribute_buffer));
+	new_accessor.get().data = std::move(vertex_attribute_buffer);
+}
+
+utki::shared_ref<accessor> gltf_loader::read_accessor(const jsondom::value& accessor_json)
+{
+	accessor::type type_v = accessor::type::vec3;
+	const std::string type_s = read_string(accessor_json, "type");
+
+	if (type_s == "SCALAR")
+		type_v = accessor::type::scalar;
+	else if (type_s == "VEC2")
+		type_v = accessor::type::vec2;
+	// else if(type_s == "VEC3") type_ = accessor::type::vec3;  // huge performance boost
```

Comment:
Could you remind what was the performance problem with VEC3?

Reply:
It was sort of a joke, because vec3 is the most probable type and it is set as default, so we don't need to check for it. 

Removed this confusing comment